torch/csrc/jit/ir.cpp

```diff
@@ -875,6 +875,197 @@ Node* Node::insertAfter(Node * n) {
   return this;
 }
 
+bool Node::moveAfterTopologicallyValid(Node* n) {
+  return tryMove(n, MoveSide::AFTER);
+}
+
+bool Node::moveBeforeTopologicallyValid(Node* n) {
+  return tryMove(n, MoveSide::BEFORE);
+}
+
+// Helper for topologically-safe node moves. See `tryMove()` for details.
+namespace {
+struct WorkingSet {
+ public:
+  explicit WorkingSet(Node* mover) {
+    add(mover);
+  }
+
+  // Add `n` to the working set
+  void add(Node* n) {
+    nodes_.push_back(n);
+    for (const auto user : getUsersSameBlock(n)) {
+      users_.insert(user);
+    }
+  }
+
+  void eraseMover() {
+    nodes_.pop_front();
+  }
+
+  const std::list<Node*>& nodes() {
+    return nodes_;
+  }
+
+  // Does the working set depend on `n`?
+  bool dependsOn(Node* n) const {
+    if (nodes_.empty()) {
+      return false;
+    }
+
+    if (n->isAfter(nodes_.front())) {
+      return producesFor(n);
+    } else {
+      return consumesFrom(n);
+    }
+  }
+
+  // Does the working set produce any values consumed by `n`?
+  bool producesFor(Node* n) const {
+    // This equivalent to asking: does the total use-set of all the nodes in the
+    // working set include `n`?
+    return users_.count(n) != 0;
+  }
+
+  // Does the working set consume any values produced by `n`?
+  bool consumesFrom(Node* n) const {
+    const auto& users = getUsersSameBlock(n);
+    return std::any_of(nodes_.begin(), nodes_.end(), [&](Node* node) {
+      return users.count(node) != 0;
+    });
+  }
+
+ private:
+  // Get all users of outputs of `n`, in the same block as `n`.
+  // This means if there is an `if` node that uses an output of `n` in some
+  // inner sub-block, we will consider the whole `if` node a user of `n`.
+  std::unordered_set<Node*> getUsersSameBlock(Node* n) const {
+    std::unordered_set<Node*> users;
+    for (const auto output : n->outputs()) {
+      for (const auto& use : output->uses()) {
+        if (use.user->owningBlock() == n->owningBlock()) {
+          users.insert(use.user);
+        } else {
+          // This user is in a sub-block. Traverse the blockchain upward until
+          // we arrive at a node that shares a block with `this`
+          auto curNode = use.user;
+          while (curNode->owningBlock() != n->owningBlock()) {
+            curNode = curNode->owningBlock()->owningNode();
+          }
```

Comment:
`JIT_ASSERT(curNode)`. In a topological messed up graph this will segfault. We should assert instead.